### DISCUSSION TITLE
iso8601 date parser for winnow

### DISCUSSION THREAD

**Discussion by soulstompp** [winnow-rs/winnow#633] [Show and tell]
I have been using the [iso8601 crate](https://crates.io/crates/iso8601) since it was fairly new, and missed it terribly when moving to winnow. So I forked it and swapped out nom for winnow, and should be fairly trivial to move existing code over from the nom version. I've published the crate as [winnow_iso8601](https://crates.io/crates/winnow_iso8601).

I tried to keep bindings as close to the stock winnow parsers as I could, and so assume this should work with any input those do. If anyone finds otherwise, please open an issue so I can fix this. 

Since I am coming from a nom mindset, I welcome feedback anywhere that I am not following winnow best practices.

**Comment by epage** [comment]
> I have been using the [iso8601 crate](https://crates.io/crates/iso8601) since it was fairly new, and missed it terribly when moving to winnow. So I forked it and swapped out nom for winnow, and should be fairly trivial to move existing code over from the nom version. I've published the crate as [winnow_iso8601](https://crates.io/crates/winnow_iso8601).

Great work!

I wonder if we should advertise integration crates like this.

> Since I am coming from a nom mindset, I welcome feedback anywhere that I am not following winnow best practices.

Maybe I'm biased but I feel like the mindset isn't all that different.  For most people, writing parsers is just writing functions.  We just have a few more combinators.  Only when you get into stream types are things weird, with checkpoints and stuff.

https://github.com/soulstompp/winnow-iso8601/blob/9aa31c3529a9709f03c066b26ff7ade701268378/src/parsers.rs#L92

Is there a reason you are doing `literal("-")` over `'-'`?

https://github.com/soulstompp/winnow-iso8601/blob/9aa31c3529a9709f03c066b26ff7ade701268378/src/parsers.rs#L40

Is there a reason you are doing `c.as_char().is_digit(10)` instead of `c.is_dec_digit()`

https://github.com/soulstompp/winnow-iso8601/blob/9aa31c3529a9709f03c066b26ff7ade701268378/src/parsers.rs#L200-L207

Can't you `seq` directly into `Date::YMD`?

https://github.com/soulstompp/winnow-iso8601/blob/9aa31c3529a9709f03c066b26ff7ade701268378/src/parsers.rs#L649-L660

Why `preceded(seq!, seq!)` rather than `seq!` with `_:`?

**Reply by soulstompp** [reply]
Thanks for the feedback! I think all of those will be good changes to add. I wasn't aware of is_dec_digit(), which seems like it is definitely what I wanted. For the last two, I think I came from a map in the original code and got stuck thinking in that pattern. I'll get all of these worked in shortly. 

I think it would be great to have these integration crates listed somewhere, I was honestly a little afraid someone was going to reply there was already a crate that filled this need for winnow users.

**Comment by soulstompp** [comment]
While working through the suggestions mentioned, I remembered why the [sign parser](https://github.com/soulstompp/winnow-iso8601/blob/9aa31c3529a9709f03c066b26ff7ade701268378/src/parsers.rs#L92) isn't just passing literal a char. Input is ` Compare<&'i str>` in this parser, and `char` doesn't have anything implemented for `char`. Switching to `Compare<char>` causes problems for any of its callers, since their Input doesn't have `Compare<char>`.  

The other fixes are in version 0.3.0, which I just released, and also has a feature for converting to time-rs and fixed a few issues around durations.

**Reply by epage** [reply]
If not a `char` literal, can't you do a `&str` literal?
